### ISSUE TITLE
EFetch parses incorrectly if ArticleTitle contains HTML-like tags

### ISSUE THREAD

**Issue by wohnlice** [biopython/biopython#1539]
My system is

> python 2.7.9 (default, Dec 10 2014, 12:28:03) [MSC v.1500 64 bit (AMD64)]
> CPython
> Windows-7-6.1.7601-SP1
> 1.70 (also tested with 1.69)

Efetch results are truncated if the article title includes HTML-like tags. This can be reproduced with PubMed ID 27797938. As you can see in XML export https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id=27797938&retmode=xml, `<ArticleTitle>` includes `<i>` tags in it, and the result is that everything before the `<i>` tag is stripped out.

I am not sure if this is a bug with Biopython or if it's an invalid entry on their part, given the DTD.

**Comment by peterjc** [comment]
Can you show a short bit of Python code illustrating the problem? Is it in the ``Bio.Entrez`` XML parser?

**Comment by wohnlice** [comment]
```
>>> from Bio import Entrez
>>> handle = Entrez.efetch(db="pubmed", id=27797938, retmode="xml")
>>> data = Entrez.read(handle)
>>> data['PubmedArticle'][0]['MedlineCitation']['Article']['ArticleTitle']
'TERTgene region and risk of pancreatic cancer.'
```

The actual value should be `Leucocyte telomere length, genetic variants at the<i>TERT</i>gene region and risk of pancreatic cancer.`

**Comment by peterjc** [comment]
Thank you,

```python
from Bio import Entrez
with open("27797938.xml", "w") as handle:
    handle.write(Entrez.efetch(db="pubmed", id=27797938, retmode="xml").read())
```

Then looking at the raw XML, we have:

```
<?xml version="1.0" ?>
<!DOCTYPE PubmedArticleSet PUBLIC "-//NLM//DTD PubMedArticle, 1st January 2018//EN" "https://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_180101.dtd">
<PubmedArticleSet>
<PubmedArticle>
...
            <ArticleTitle>Leucocyte telomere length, genetic variants at the<i>TERT</i>gene region and risk of pancreatic cancer.</ArticleTitle>
...
```

Perhaps those italics HTML tags shouldn't be there, but I think you're right about Biopython doing something odd.

CC @mdehoon

**Comment by wohnlice** [comment]
Received this response from NLM

> As of last October, we accept &lt;b&gt;, &lt;i&gt;, &lt;sup&gt;, &lt;sub&gt;, and &lt;u&gt; as valid XML tags. These elements are defined in the PubMed 2018 DTD: https://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_180101.dtd and should be treated like all of the other elements in the DTD (not as an HTML tag in this context). Please see also <ArticleTitle></ArticleTitle> in the DTD for clarification of the content model for this element

**Comment by peterjc** [comment]
I'm not quite sure what they mean about the tag treatment, but that might indirectly explain the apparent lack of spaces before/after the italics word.

**Comment by reiaoki** [comment]
I am having the similar issues for `<sup>` and `<sub> `tags. It looks like the problem occurs when the _read_ function runs.  I can see the full text of abstract when I just run _efetch_. 
```
handle = Entrez.efetch(db = 'pubmed', retmode = 'xml', id = '29106400')
print(handle.readlines())
```
The above code gave me the whole xml results from the Eutils. 

But when I run the read function. I see my text is truncated (here, I extracted out the abstract part). 
```
handle = Entrez.efetch(db = 'pubmed', retmode = 'xml', id = '29106400')
fetchResults = Entrez.read(handle)
print(fetchResults['PubmedArticle'][0]['MedlineCitation']['Article']['Abstract'])
```
will return 

` {'AbstractText': ['+lymphocytes. T cells transduced with MMG49-derived chimeric antigen receptor (CAR) exerted anti-MM effects without damaging normal hematopoietic cells. Thus, MMG49 CAR T cell therapy is promising for MM, and a receptor protein with a rare but physiologically relevant conformation can serve as a cancer immunotherapy target.'], 'sub': '7', 'sup': '+'}`

I am guessing the `'sub': '7', 'sup': '+'` part at the end has something to do with the problem (biopython unrecognizable tag?). Any thoughts? It is definitely the XML parsing from the biopython that is causing the issue I think.

**Comment by reiaoki** [comment]
@peterjc Any thoughts on how to fix `Entrez.parse` or `Entrez.read` functions?

**Comment by mdehoon** [comment]
I think the problem occurs in Bio/Entrez/Parser.py in startElementHandler, where self.content is set to "" (and therefore any existing data in self.content is dropped) whenever a string-like tag is seen in the XML file.
Previously, the parser would see
<ArticleTitle>My title</ArticleTitle>
and the steps it goes through were
<ArticleTitle>: set self.content to ""
My title: add "My title" to self.content
</ArticleTitle>: store self.content in 'ArticleTitle'

Now, the parser may see
<ArticleTitle>My <i>italic</i> title</ArticleTitle>
and the steps it goes through are
<ArticleTitle>: set self.content to ""
My : add "My " to self.content
<i>: set self.content to ""
italic: add "italic" to self.content
</i>: store self.content in "ArticleTitle"
 title: add " title" to self.content
</ArticleTitle>: store self.content in "ArticleTitle"

so anything before <i> is lost.

To solve this bug, we need to figure out some way for the parser to decide if a tag should reset self.content to "".

@reiaoki  Does this help you to debug this?
</ArticleTitle>: store self.content in 'ArticleTitle'

**Comment by peterjc** [comment]
@mdehoon I had a little experiment with this, and it seems hard. I wonder if a another option to is to special case the ``<i>``, ``<b>``, ``<sup>`` etc tags to be treated as literals within a string? I don't understand the parser well enough to know how to try that though.

**Comment by OscarMaestre** [comment]
Sorry but another patch I uploaded references incorrectly this issue. In fact, that patch doesn't have anything to do with this issue. Sorry again

**Comment by OscarMaestre** [comment]
I studied the way the parser works and the addition of <i> (<u><b> etc...) seem to break completely the way the parser works. I can create a small patch that solves this problems but ignoring such tags. Also I think that rewrite the parser can be a really complex task. The class DataHandler seems to have the ability to parse the DTD (really complex) and based on that be able to recognise the different elements that can appear in the XML file.

I also think that the main problem is that now the DTD allows tags to appear in the middle of another tag which breaks the way the parser works.

**Comment by peterjc** [comment]
I think Ignoring text formatting tags ``<i>``, ``<b>``, ``<u>``, ``<sup>``, etc within titles would be an acceptable fix.

If you don't care about the DTD, then Python provides multiple XML parsers which can be used directly instead.

**Comment by peterjc** [comment]
Oscar's pull request is a bit mixed up, but the core idea is simple:

```python
        if name not in ("i", "u", "b", "sup", "sub"):
            self.content = ""
```

i.e. Hard code when not to reset the content string for these special cases.

**Comment by OscarMaestre** [comment]
Sorry again. I will clean the code and submit another pull request.

**Comment by wohnlice** [comment]
Oscar, this seems to strip some whitespace adjacent to the tags and this is also incorrectly reflected in the unit test. You are testing for "theTERTgene" but it should really be "the TERT gene"

**Comment by wohnlice** [comment]
I see. That's a bit strange since the PubMed entry has spaces https://www.ncbi.nlm.nih.gov/pubmed/27797938. Would it make sense and/or be possible to assume any tags here should imply whitespace?

**Comment by OscarMaestre** [comment]
Yes, in fact it's easy, but I'm not sure if that's the way these tags should be treated.

**Comment by peterjc** [comment]
You're both right, the original example was ``Leucocyte telomere length, genetic variants at the<i>TERT</i>gene region and risk of pancreatic cancer`` so if you take that at face value there are no spaces either side of the word TERT - which is the NCBI's fault. I wonder if the HTML standards cover this and say there is implicitly white space there?

**Comment by MarkusPiotrowski** [comment]
I doubt that this is HTML standard...
I just want to point to some problems. At first glance I thought it wouldn't harm just to add whitespaces around tagged words, but then I realized that for some tags you wouldn't like whitespaces on both sides (super- and subscript) and then it may not even be clear where a whitespace may be sitting, if any:
<sup>3</sup>H, H<sub>2</sub>O, &beta;<sub>7</sub>. 

How does NLM handle this? If there are no whitespaces in the xml file how do they know where to put them?

**Comment by MarkusPiotrowski** [comment]
Extending on my former comment: There seems to be no sensible way for the parser to know where it should place whitespaces when removing a format tag. 

So I (without knowing the internals of the parser) would suggest 'just' to add these tags to the content, instead of ignoring them. So instead of:
```python
if name not in ("i", "u", "b", "sup", "sub"):
    self.content = ""
```

do: 
```python
if name in ["i", "u", "b", "sup", "sub"]:
    self.content += '<%s>' % name
else:
    self.content = ""
```

And do something similar in ```endElementHandler```:
```python
...
if name in ["i", "u", "b", "sup", "sub"]:
    self.content += '</%s>' % name
```

This way the content is exactly that what the xml file contains.

```python
>>> from Bio import Entrez
>>> handle = Entrez.efetch(db="pubmed", id=27797938, retmode="xml")
>>> data = Entrez.read(handle)
>>> data['PubmedArticle'][0]['MedlineCitation']['Article']['ArticleTitle']
'Leucocyte telomere length, genetic variants at the<i>TERT</i>gene region and risk of pancreatic cancer.'
```

**Comment by peterjc** [comment]
Markus' idea would preserve the raw HTML of the title as is, missing spaces and HTML tags and all. I think this would be a neat way to deal with this situation.

**Comment by mdehoon** [comment]
@peterjc  , @MarkusPiotrowski  I agree with Markus' suggestion to keep the raw HTML in the title.

**Comment by mdehoon** [comment]
Note btw that the DTD specifies which HTML tags can appear as part of the text, so it should not be necessary to hard-code this in the parser. The relevant code is in lines 487-491 of Parser.py. At that point, by looking at the contents of the model variable, you could figure out which HTML tags can appear in the text.

**Comment by OscarMaestre** [comment]
Ok, no problemo :) Let's preserve original HTML in the original ArticleTitle element by hard-coding i,u,b,sup and sub tags

**Comment by peterjc** [comment]
Michael says we don't need to hard code this - https://github.com/biopython/biopython/issues/1539#issuecomment-378942964

**Comment by OscarMaestre** [comment]
Oh, I see, thank you

**Comment by OscarMaestre** [comment]
I was reading Parser.py but I'm not able to understand the intentions of the code, so I'm not able to imagine how to solve this issue. I keep on reading :)

**Comment by peterjc** [comment]
Adding a little debugging, at the point we need to clear ``self.content`` (or not),

```python
self.strings = [u'AbstractText', u'AccessionNumber', u'Acronym', u'Affiliation', u'Agency', u'ArticleId', u'ArticleTitle', u'b', u'BookTitle', u'CitationString', u'CitationSubset', u'CoiStatement', u'CollectionTitle', u'CollectiveName', u'ContractNumber', u'CopyrightInformation', u'Country', u'DataBankName', u'Day', u'DescriptorName', u'Edition', u'ELocationID', u'EndPage', u'ForeName', u'GeneSymbol', u'GeneralNote', u'GrantID', u'Hour', u'i', u'Identifier', u'Initials', u'Isbn', u'ISOAbbreviation', u'ISSN', u'ISSNLinking', u'Issue', u'Item', u'Keyword', u'Language', u'LastName', u'LocationLabel', u'Medium', u'MedlineDate', u'MedlinePgn', u'MedlineTA', u'Minute', u'Month', u'NameOfSubstance', u'NlmUniqueID', u'Note', u'NumberOfReferences', u'OtherID', u'PMID', u'Param', u'PublicationStatus', u'PublicationType', u'PublisherLocation', u'PublisherName', u'QualifierName', u'RefSource', u'RegistryNumber', u'ReportNumber', u'Season', u'Second', u'SectionTitle', u'SpaceFlightMission', u'StartPage', u'sub', u'Suffix', u'sup', u'SupplMeshName', u'Title', u'u', u'URL', u'VernacularTitle', u'Volume', u'VolumeTitle', u'Year']
```

The leading u is for unicode since I ran this on Python 2:

That list includes ``b``, ``i``, ``u``, ``sup`` and ``sub`` along with lots of others things including ``Title``.

i.e. This list alone does not help.

**Comment by peterjc** [comment]
Excerpts from the DTD file, https://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_180101.dtd

```xml
...
<!ENTITY % text             "(#PCDATA | b | i | sup | sub | u)*" >
...
<!ELEMENT	b		%text; > <!-- bold -->
...
<!ELEMENT	i		%text; > <!-- italic -->
...
<!ELEMENT	sub	%text; > <!-- subscript -->

<!ELEMENT	Suffix %text;>

<!ELEMENT	sup	%text; > <!-- superscript -->

<!ELEMENT	SupplMeshList (SupplMeshName+)>

<!ELEMENT	SupplMeshName (#PCDATA) >
<!ATTLIST	SupplMeshName 
		    Type (Disease | Protocol | Organism) #REQUIRED
		    UI CDATA #REQUIRED >
		    
<!ELEMENT	Title (#PCDATA) >

<!ELEMENT	u		%text; > <!-- underline -->  
...
```

It looks like the distinction we need to look at is: Is the XML tag labelled as ``#PCDATA`` (e.g. title) or not (``b``, ``i``, ``u``, ``sup`` and ``sub``).

**Comment by peterjc** [comment]
See also https://www.ncbi.nlm.nih.gov/mailman/pipermail/utilities-announce/2018-March/000125.html

The new DTD may alter things like this?

**Comment by peterjc** [comment]
@wohnlice @OscarMaestre are you able to retest with the fix from @mdehoon (on pull request #1640, now merged into the master branch)? This turned out to be a much bigger change than the approach being tried in #1597. Thanks.

**Comment by mdehoon** [comment]
As far as I can tell, this has been fixed by pull request #1640 . Please reopen this issue if the error still appears.
